transports/websocket-websys/src/lib.rs

```diff
@@ -0,0 +1,445 @@
+            }
+        }
+    }
+
+    fn error_barrier(&self) -> io::Result<()> {
+        if self.errored.load(Ordering::SeqCst) {
+            return Err(io::ErrorKind::BrokenPipe.into());
+        }
+
+        Ok(())
+    }
+}
+
+/// The state of the WebSocket.
+///
+/// See <https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/readyState>.
+#[derive(PartialEq)]
+enum ReadyState {
+    Connecting,
+    Open,
+    Closing,
+    Closed,
+}
+
+impl Connection {
+    fn new(socket: WebSocket) -> Self {
+        socket.set_binary_type(web_sys::BinaryType::Arraybuffer);
+
+        let open_waker = Rc::new(AtomicWaker::new());
+        let onopen_closure = Closure::<dyn FnMut(_)>::new({
+            let open_waker = open_waker.clone();
+            move |_| {
+                open_waker.wake();
+            }
+        });
+        socket.set_onopen(Some(onopen_closure.as_ref().unchecked_ref()));
+
+        let close_waker = Rc::new(AtomicWaker::new());
+        let onclose_closure = Closure::<dyn FnMut(_)>::new({
+            let close_waker = close_waker.clone();
+            move |_| {
+                close_waker.wake();
+            }
+        });
+        socket.set_onclose(Some(onclose_closure.as_ref().unchecked_ref()));
+
+        let errored = Rc::new(AtomicBool::new(false));
+        let onerror_closure = Closure::<dyn FnMut(_)>::new({
+            let errored = errored.clone();
+            move |_| {
+                errored.store(true, Ordering::SeqCst);
+            }
+        });
+        socket.set_onerror(Some(onerror_closure.as_ref().unchecked_ref()));
+
+        let read_buffer = Rc::new(Mutex::new(BytesMut::new()));
+        let new_data_waker = Rc::new(AtomicWaker::new());
+        let onmessage_closure = Closure::<dyn FnMut(_)>::new({
+            let read_buffer = read_buffer.clone();
+            let new_data_waker = new_data_waker.clone();
+            let errored = errored.clone();
+            move |e: MessageEvent| {
+                let data = js_sys::Uint8Array::new(&e.data());
+
+                let mut read_buffer = read_buffer.lock().unwrap();
+
+                if read_buffer.len() + data.length() as usize > MAX_BUFFER {
+                    log::warn!("Remote is overloading us with messages, closing connection");
+                    errored.store(true, Ordering::SeqCst);
+
+                    return;
+                }
+
+                read_buffer.extend_from_slice(&data.to_vec());
+                new_data_waker.wake();
+            }
+        });
+        socket.set_onmessage(Some(onmessage_closure.as_ref().unchecked_ref()));
+
+        let write_waker = Rc::new(AtomicWaker::new());
+        let on_buffered_amount_low_closure = Closure::<dyn FnMut(_)>::new({
+            let write_waker = write_waker.clone();
+            let socket = socket.clone();
+            move |_| {
+                if socket.buffered_amount() == 0 {
+                    write_waker.wake();
+                }
```

Comment:
Why only wake the `write_waker` when `buffered_amount` reaches `0`. Why not earlier, e.g. `MAX_BUFFER / 2`?

Reply:
No particular reason, happy to change it to `MAX_BUFFER / 2`! :)

Reply:
I am using the same waker for flushing. If we wake at `MAX_BUFFER / 2`, we'd need two wakers to avoid unnecessary wake-ups, right? Should we do that or leave it at `== 0`?

Reply:
Fine as is. Though I suggest lowering the interval below.